AppTaskQueue/appscale/taskqueue/queue.py

```diff
@@ -235,7 +235,538 @@ def __repr__(self):
     return '<PushQueue {}: {}>'.format(self.name, attr_str)
 
 
+class PostgresPullQueue(Queue):
+
+  """
+  Before using Postgres implementation, make sure that
+  connection using appscale user can be created:
+  /etc/postgresql/9.5/main/pg_hba.conf
+  """
+
+  def __init__(self, queue_info, app, pg_connection=None):
+    """ Create a PostgresPullQueue object.
+
+    Args:
+      queue_info: A dictionary containing queue info.
+      app: A string containing the application ID.
+      pg_connection: A psycopg2 connection to PostgreSQL.
+    """
+    super(PostgresPullQueue, self).__init__(queue_info, app)
+    self.pg_connection = pg_connection
+    self.pg_cursor = pg_connection.cursor()
+    self.ensure_tables_created()
+
+  @property
+  def tasks_table_name(self):
+    return 'pullqueue-{}'.format(self.name)
+
+  def ensure_tables_created(self):
+    try:
+      self.pg_cursor.execute(
+        'CREATE TABLE IF NOT EXISTS "{table_name}" ('
+        '  task_name varchar(500) NOT NULL,'
+        '  time_enqueued timestamp NOT NULL,'
+        '  retry_count integer NOT NULL,'
+        '  lease_expires timestamp NOT NULL,'
+        '  payload text,'
+        '  tag varchar(500),'
+        '  PRIMARY KEY (task_name)'
+        ');'
+        'CREATE INDEX "{table_name}-eta-retry-tag-index" '
+        '  ON "{table_name}" USING BTREE (lease_expires, retry_count, tag);'
+        'CREATE INDEX "{table_name}-retry-eta-tag-index" '
+        '  ON "{table_name}" (retry_count, lease_expires, tag);'
+        .format(table_name=self.tasks_table_name)
+      )
+      self.pg_connection.commit()
+    except Exception as err:
+      logger.error('Rolling back transaction ({err})'.format(err=err))
+      self.pg_connection.rollback()
+      raise
+
+  def add_task(self, task):
+    """ Adds a task to the queue.
+
+    Args:
+      task: A Task object.
+    Raises:
+      InvalidTaskInfo if the task ID already exists in the queue
+        or it doesn't have payloadBase64 attribute.
+    """
+    import psycopg2
+    if not hasattr(task, 'payloadBase64'):
+      raise InvalidTaskInfo('{} is missing a payload.'.format(task))
+
+    enqueue_time = datetime.datetime.utcnow()
+    try:
+      lease_expires = task.leaseTimestamp
+    except AttributeError:
+      lease_expires = datetime.datetime.utcfromtimestamp(0)
```

Comment:
Since Postgres does not suffer from the same clock issues that can affect immediate availabilty on Cassandra, this can use the current timestamp (also generated by Postgres). That allows leases to be ordered by enqueue time when eta is not specified.